File: internal/permissions/permissions.go
// Package permissions defines all permission constants and types for the PBAC system.
package permissions

import "strings"

// Permission represents a permission code string.
type Permission string

// Category represents the category of a permission.
type Category string

const (
	CategorySystem Category = "system"
	CategoryUI     Category = "ui"
	CategoryRCON   Category = "rcon"
)

// Wildcard permission grants all access.
const Wildcard Permission = "*"

// UI Permissions - Control access to pages/components.
const (
	UIDashboardView   Permission = "ui:dashboard:view"
	UIAuditLogsView   Permission = "ui:audit_logs:view"
	UIMetricsView     Permission = "ui:metrics:view"
	UIFeedsView       Permission = "ui:feeds:view"
	UIConsoleView     Permission = "ui:console:view"
	UIConsoleExecute  Permission = "ui:console:execute"
	UIPluginsView     Permission = "ui:plugins:view"
	UIPluginsManage   Permission = "ui:plugins:manage"
	UIWorkflowsView   Permission = "ui:workflows:view"
	UIWorkflowsManage Permission = "ui:workflows:manage"
	UISettingsView    Permission = "ui:settings:view"
	UISettingsManage  Permission = "ui:settings:manage"
	UIUsersManage     Permission = "ui:users:manage"
	UIRolesManage     Permission = "ui:roles:manage"
	UIBansView        Permission = "ui:bans:view"
	UIBansCreate      Permission = "ui:bans:create"
	UIBansEdit        Permission = "ui:bans:edit"
	UIBansDelete      Permission = "ui:bans:delete"
	UIPlayersView     Permission = "ui:players:view"
	UIPlayersKick     Permission = "ui:players:kick"
	UIPlayersWarn     Permission = "ui:players:warn"
	UIPlayersMove     Permission = "ui:players:move"
	UIRulesView       Permission = "ui:rules:view"
	UIRulesManage     Permission = "ui:rules:manage"
	UIBanListsView    Permission = "ui:ban_lists:view"
	UIBanListsManage  Permission = "ui:ban_lists:manage"
	UIMOTDView        Permission = "ui:motd:view"
	UIMOTDManage      Permission = "ui:motd:manage"
)

// RCON/Squad Permissions - Map to Squad's admin.cfg permissions.
const (
	RCONReserve         Permission = "rcon:reserve"
	RCONBalance         Permission = "rcon:balance"
	RCONCanSeeAdminChat Permission = "rcon:canseeadminchat"
	RCONManageServer    Permission = "rcon:manageserver"
	RCONTeamChange      Permission = "rcon:teamchange"
	RCONChat            Permission = "rcon:chat"
	RCONCameraman       Permission = "rcon:cameraman"
	RCONKick            Permission = "rcon:kick"
	RCONBan             Permission = "rcon:ban"
	RCONForceTeamChange Permission = "rcon:forceteamchange"
	RCONImmune          Permission = "rcon:immune"
	RCONChangeMap       Permission = "rcon:changemap"
	RCONPause           Permission = "rcon:pause"
	RCONCheat           Permission = "rcon:cheat"
	RCONPrivate         Permission = "rcon:private"
	RCONConfig          Permission = "rcon:config"
	RCONFeatureTest     Permission = "rcon:featuretest"
	RCONDemos           Permission = "rcon:demos"
	RCONDisbandSquad    Permission = "rcon:disbandsquad"
	RCONRemoveFromSquad Permission = "rcon:removefromsquad"
	RCONDemoteCommander Permission = "rcon:demotecommander"
	RCONDebug           Permission = "rcon:debug"
)

// SquadPermissionMap maps RCON permission codes to Squad's admin.cfg format.
var SquadPermissionMap = map[Permission]string{
	RCONReserve:         "reserve",
	RCONBalance:         "balance",
	RCONCanSeeAdminChat: "canseeadminchat",
	RCONManageServer:    "manageserver",
	RCONTeamChange:      "teamchange",
	RCONChat:            "chat",
	RCONCameraman:       "cameraman",
	RCONKick:            "kick",
	RCONBan:             "ban",
	RCONForceTeamChange: "forceteamchange",
	RCONImmune:          "immune",
	RCONChangeMap:       "changemap",
	RCONPause:           "pause",
	RCONCheat:           "cheat",
	RCONPrivate:         "private",
	RCONConfig:          "config",
	RCONFeatureTest:     "featuretest",
	RCONDemos:           "demos",
	RCONDisbandSquad:    "disbandSquad",
	RCONRemoveFromSquad: "removeFromSquad",
	RCONDemoteCommander: "demoteCommander",
	RCONDebug:           "debug",
}

// ReverseSquadPermissionMap maps Squad admin.cfg permission names to RCON permission codes.
var ReverseSquadPermissionMap = map[string]Permission{
	"reserve":         RCONReserve,
	"balance":         RCONBalance,
	"canseeadminchat": RCONCanSeeAdminChat,
	"manageserver":    RCONManageServer,
	"teamchange":      RCONTeamChange,
	"chat":            RCONChat,
	"cameraman":       RCONCameraman,
	"kick":            RCONKick,
	"ban":             RCONBan,
	"forceteamchange": RCONForceTeamChange,
	"immune":          RCONImmune,
	"changemap":       RCONChangeMap,
	"pause":           RCONPause,
	"cheat":           RCONCheat,
	"private":         RCONPrivate,
	"config":          RCONConfig,
	"featuretest":     RCONFeatureTest,
	"demos":           RCONDemos,
	"disbandSquad":    RCONDisbandSquad,
	"removeFromSquad": RCONRemoveFromSquad,
	"demoteCommander": RCONDemoteCommander,
	"debug":           RCONDebug,
}

// GetCategory returns the category of a permission.
func (p Permission) GetCategory() Category {
	if p == Wildcard {
		return CategorySystem
	}

	parts := strings.SplitN(string(p), ":", 2)
	if len(parts) < 2 {
		return CategorySystem
	}

	switch parts[0] {
	case "ui":
		return CategoryUI
	case "rcon":
		return CategoryRCON
	default:
		return CategorySystem
	}
}

// ToSquadPermission converts an RCON permission to its Squad admin.cfg equivalent.
// Returns empty string if not an RCON permission.
func (p Permission) ToSquadPermission() string {
	if squadPerm, ok := SquadPermissionMap[p]; ok {
		return squadPerm
	}
	return ""
}

// IsRCON returns true if this is an RCON permission.
func (p Permission) IsRCON() bool {
	return p.GetCategory() == CategoryRCON
}

// IsUI returns true if this is a UI permission.
func (p Permission) IsUI() bool {
	return p.GetCategory() == CategoryUI
}

// IsWildcard returns true if this is the wildcard permission.
func (p Permission) IsWildcard() bool {
	return p == Wildcard
}

// String returns the string representation of the permission.
func (p Permission) String() string {
	return string(p)
}

// AllPermissions returns all defined permissions.
func AllPermissions() []Permission {
	return []Permission{
		Wildcard,
		// UI
		UIDashboardView, UIAuditLogsView, UIMetricsView, UIFeedsView,
		UIConsoleView, UIConsoleExecute, UIPluginsView, UIPluginsManage,
		UIWorkflowsView, UIWorkflowsManage, UISettingsView, UISettingsManage,
		UIUsersManage, UIRolesManage, UIBansView, UIBansCreate, UIBansEdit, UIBansDelete,
		UIPlayersView, UIPlayersKick, UIPlayersWarn, UIPlayersMove,
		UIRulesView, UIRulesManage, UIBanListsView, UIBanListsManage,
		UIMOTDView, UIMOTDManage,
		// RCON
		RCONReserve, RCONBalance, RCONCanSeeAdminChat, RCONManageServer,
		RCONTeamChange, RCONChat, RCONCameraman, RCONKick, RCONBan,
		RCONForceTeamChange, RCONImmune, RCONChangeMap, RCONPause,
		RCONCheat, RCONPrivate, RCONConfig, RCONFeatureTest, RCONDemos,
		RCONDisbandSquad, RCONRemoveFromSquad, RCONDemoteCommander, RCONDebug,
	}
}

// UIPermissions returns all UI permissions.
func UIPermissions() []Permission {
	return []Permission{
		UIDashboardView, UIAuditLogsView, UIMetricsView, UIFeedsView,
		UIConsoleView, UIConsoleExecute, UIPluginsView, UIPluginsManage,
		UIWorkflowsView, UIWorkflowsManage, UISettingsView, UISettingsManage,
		UIUsersManage, UIRolesManage, UIBansView, UIBansCreate, UIBansEdit, UIBansDelete,
		UIPlayersView, UIPlayersKick, UIPlayersWarn, UIPlayersMove,
		UIRulesView, UIRulesManage, UIBanListsView, UIBanListsManage,
		UIMOTDView, UIMOTDManage,
	}
}

// RCONPermissions returns all RCON permissions.
func RCONPermissions() []Permission {
	return []Permission{
		RCONReserve, RCONBalance, RCONCanSeeAdminChat, RCONManageServer,
		RCONTeamChange, RCONChat, RCONCameraman, RCONKick, RCONBan,
		RCONForceTeamChange, RCONImmune, RCONChangeMap, RCONPause,
		RCONCheat, RCONPrivate, RCONConfig, RCONFeatureTest, RCONDemos,
		RCONDisbandSquad, RCONRemoveFromSquad, RCONDemoteCommander, RCONDebug,
	}
}

File: web/constants/permissions.ts
// Permission constants for the PBAC system
// These must stay in sync with internal/permissions/permissions.go

export type PermissionCategory = "system" | "ui" | "rcon";

// Wildcard permission grants all access
export const PERMISSION_WILDCARD = "*";

// UI Permissions - Control access to pages/components
export const UI_PERMISSIONS = {
  DASHBOARD_VIEW: "ui:dashboard:view",
  AUDIT_LOGS_VIEW: "ui:audit_logs:view",
  METRICS_VIEW: "ui:metrics:view",
  FEEDS_VIEW: "ui:feeds:view",
  CONSOLE_VIEW: "ui:console:view",
  CONSOLE_EXECUTE: "ui:console:execute",
  PLUGINS_VIEW: "ui:plugins:view",
  PLUGINS_MANAGE: "ui:plugins:manage",
  WORKFLOWS_VIEW: "ui:workflows:view",
  WORKFLOWS_MANAGE: "ui:workflows:manage",
  SETTINGS_VIEW: "ui:settings:view",
  SETTINGS_MANAGE: "ui:settings:manage",
  USERS_MANAGE: "ui:users:manage",
  ROLES_MANAGE: "ui:roles:manage",
  BANS_VIEW: "ui:bans:view",
  BANS_CREATE: "ui:bans:create",
  BANS_EDIT: "ui:bans:edit",
  BANS_DELETE: "ui:bans:delete",
  PLAYERS_VIEW: "ui:players:view",
  PLAYERS_KICK: "ui:players:kick",
  PLAYERS_WARN: "ui:players:warn",
  PLAYERS_MOVE: "ui:players:move",
  RULES_VIEW: "ui:rules:view",
  RULES_MANAGE: "ui:rules:manage",
  BAN_LISTS_VIEW: "ui:ban_lists:view",
  BAN_LISTS_MANAGE: "ui:ban_lists:manage",
  MOTD_VIEW: "ui:motd:view",
  MOTD_MANAGE: "ui:motd:manage",
} as const;

// RCON/Squad Permissions - Map to Squad's admin.cfg permissions
export const RCON_PERMISSIONS = {
  RESERVE: "rcon:reserve",
  BALANCE: "rcon:balance",
  CAN_SEE_ADMIN_CHAT: "rcon:canseeadminchat",
  MANAGE_SERVER: "rcon:manageserver",
  TEAM_CHANGE: "rcon:teamchange",
  CHAT: "rcon:chat",
  CAMERAMAN: "rcon:cameraman",
  KICK: "rcon:kick",
  BAN: "rcon:ban",
  FORCE_TEAM_CHANGE: "rcon:forceteamchange",
  IMMUNE: "rcon:immune",
  CHANGE_MAP: "rcon:changemap",
  PAUSE: "rcon:pause",
  CHEAT: "rcon:cheat",
  PRIVATE: "rcon:private",
  CONFIG: "rcon:config",
  FEATURE_TEST: "rcon:featuretest",
  DEMOS: "rcon:demos",
  DISBAND_SQUAD: "rcon:disbandsquad",
  REMOVE_FROM_SQUAD: "rcon:removefromsquad",
  DEMOTE_COMMANDER: "rcon:demotecommander",
  DEBUG: "rcon:debug",
} as const;

// Combined permissions object for convenience
export const Permissions = {
  WILDCARD: PERMISSION_WILDCARD,
  UI: UI_PERMISSIONS,
  RCON: RCON_PERMISSIONS,
} as const;

// Type for any permission string
export type Permission =
  | typeof PERMISSION_WILDCARD
  | (typeof UI_PERMISSIONS)[keyof typeof UI_PERMISSIONS]
  | (typeof RCON_PERMISSIONS)[keyof typeof RCON_PERMISSIONS];

// Maps RCON permission codes to Squad's admin.cfg format
export const SQUAD_PERMISSION_MAP: Record<string, string> = {
  "rcon:reserve": "reserve",
  "rcon:balance": "balance",
  "rcon:canseeadminchat": "canseeadminchat",
  "rcon:manageserver": "manageserver",
  "rcon:teamchange": "teamchange",
  "rcon:chat": "chat",
  "rcon:cameraman": "cameraman",
  "rcon:kick": "kick",
  "rcon:ban": "ban",
  "rcon:forceteamchange": "forceteamchange",
  "rcon:immune": "immune",
  "rcon:changemap": "changemap",
  "rcon:pause": "pause",
  "rcon:cheat": "cheat",
  "rcon:private": "private",
  "rcon:config": "config",
  "rcon:featuretest": "featuretest",
  "rcon:demos": "demos",
  "rcon:disbandsquad": "disbandSquad",
  "rcon:removefromsquad": "removeFromSquad",
  "rcon:demotecommander": "demoteCommander",
  "rcon:debug": "debug",
};

// Helper to get permission category
export function getPermissionCategory(permission: string): PermissionCategory {
  if (permission === PERMISSION_WILDCARD) return "system";
  const prefix = permission.split(":")[0];
  switch (prefix) {
    case "ui":
      return "ui";
    case "rcon":
      return "rcon";
    default:
      return "system";
  }
}

// Helper to check if permission is RCON type
export function isRconPermission(permission: string): boolean {
  return permission.startsWith("rcon:");
}

// Helper to check if permission is UI type
export function isUiPermission(permission: string): boolean {
  return permission.startsWith("ui:");
}

// Helper to convert RCON permission to Squad format
export function toSquadPermission(permission: string): string | null {
  return SQUAD_PERMISSION_MAP[permission] || null;
}

// Permission category display names
export const PERMISSION_CATEGORY_NAMES: Record<PermissionCategory, string> = {
  system: "System",
  ui: "User Interface",
  rcon: "RCON / Squad Server",
};

// All permissions as a flat array
export const ALL_PERMISSIONS: Permission[] = [
  PERMISSION_WILDCARD,
  ...Object.values(UI_PERMISSIONS),
  ...Object.values(RCON_PERMISSIONS),
];
